### PR TITLE
Add testing support with pytest and dpytest

### PR DESCRIPTION
**Pull request by aldehir**
Add testing support with pytest and dpytest.

Because the bot uses globals, it is difficult to isolate each test. Instead, we run a single bot and event loop for the entirety of the test run.

How to run:
```console
poetry run pytest
```

### CODE REVIEW

**Comment by aldehir** [comment]
@dcramps I added a helper to return the message as soon as there is one in the queue. Let me know what you think, although not sure I like the name `helper`. The downside is we can't use the fluent API in dpytest, but I'm not a real fan of it anyway.

File: test/conftest.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    loop = asyncio.get_running_loop()
+    background_task = loop.create_task(run_background_tasks())
+
+    yield client.c
+
+    console.terminate()
+    await background_task
+
+
+class Helper:
```
Comment (dcramps) [inline test/conftest.py]: call it messageservice or something? i hate it when things are named helper lol

File: test/conftest.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    def __init__(self, loop=None):
+        self.loop = loop or asyncio.get_running_loop()
+
+    async def message(self, peek=False, timeout=0.5) -> discord.Message:
```
Comment (dcramps) [inline test/conftest.py]: if you call it `messageservice` you could rename this to `get_message` or `retrieve_message` or something. `messageservice.get_message(...)` would read fairly nicely in test functions.

File: test/test_bot.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+import discord
+import discord.ext.test as dpytest
+
+
+@pytest.mark.asyncio(scope="session")
+async def test_enable_pickups(bot, helper):
+    await dpytest.message("!enable_pickups", member=0)
+
+    resp = await helper.message()
```
Comment (dcramps) [inline test/test_bot.py]: why resp? it's a message, so call it that. :thonk:

`message.content` is nicer than `resp.content` too

**Review by dcramps — APPROVED** [review]
a few nitpicks + i would rename conftest to fixtures or something more meaningful. otherwise lgtm, merge when ready

File: test/conftest.py (hunk 1 above)
Comment (aldehir) [reply test/conftest.py]: I agree, how about `messenger`?

**Comment by aldehir** [comment]
`conftest.py` is a convention in pytest[^1]. The pytest runner will execute conftest.py before any other test modules.

[^1]: https://docs.pytest.org/en/8.0.x/reference/fixtures.html#conftest-py-sharing-fixtures-across-multiple-files

**Pull request merged**